### PR TITLE
PTFM-13654 dxpy config manager

### PR DESCRIPTION
**Pull request by kislyuk**
@psung, please review.

* Unify dxpy configuration management
* Save and load entire configuration at once, never piecemeal
* Simplify and deduplicate code in some places
* Begin test suite

### CODE REVIEW

File: src/python/dxpy/utils/config.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+                dxpy.set_workspace_id(environ["DX_PROJECT_CONTEXT_ID"])
+                dxpy.set_project_context(environ["DX_PROJECT_CONTEXT_ID"])
+
+    def get_global_conf_dir(self):
+        return self._global_conf_dir
+
+    def get_user_conf_dir(self):
+        return self._user_conf_dir
+
+    def get_session_conf_dir(self, cleanup=True):
```
Comment (psung) [inline src/python/dxpy/utils/config.py]: Perhaps `cleanup` should default to False and only be set to True when initializing. It seems superfluous to attempt the cleanup when saving and clearing the config, too.

File: src/python/dxpy/utils/config.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        if key in self.CORE_VAR_NAMES:
+            self._sync_dxpy_state()
+
+    def __iter__(self):
+        for item in self.VAR_NAMES:
+            if item in environ:
+                yield item
+
+    def __len__(self):
+        return len(self.VAR_NAMES)
```
Comment (psung) [inline src/python/dxpy/utils/config.py]: The length reported here ought to be consistent with the number of items returned by the iterator.

File: src/python/dxpy/utils/config.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+    def _write_conf_dir(self, conf_dir):
+        try:
+            os.makedirs(conf_dir, 0o700)
+        except OSError:
+            os.chmod(conf_dir, 0o700)
+
+        env_jsonfile_path = os.path.join(conf_dir, "environment.json")
+        # Make sure the file has 600 permissions
+        with _open_for_writing_with_permissions(env_jsonfile_path, 0o600) as fd:
```
Comment (psung) [inline src/python/dxpy/utils/config.py]: An alternative strategy might be to write to a separate file and (atomically) rename it to `environment.json`, but maybe that's overkill for now.

File: src/python/dxpy/utils/config.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+            json.dump({k: self.get(k, self.defaults.get(k)) for k in self}, fd, indent=4)
+            fd.write("\n")
+
+        for var in self.STANDALONE_VAR_NAMES:
+            # Make sure the file has 600 permissions
+            with _open_for_writing_with_permissions(os.path.join(conf_dir, var), 0o600) as fd:
+                value = self.get(var, self.defaults.get(var, ""))
+                fd.write(value.encode(sys_encoding) if USING_PYTHON2 else value)
+
+        if not os.path.exists(os.path.join(self._user_conf_dir, "unsetenv")):
```
Comment (psung) [inline src/python/dxpy/utils/config.py]: I feel that this should be factored out into a separate function, perhaps to be called just once in `save`. It's independent of the values in the configuration, and it shouldn't be written twice (once for each call to `_write_conf_dir`.

File: src/python/dxpy/scripts/dx.py [hunk 5]
```diff
@@ -3446,7 +3438,7 @@ def register_subparser(subparser, subparsers_action=None, categories=('other', )
                                         description='Clears all environment variables set by dx.  More specifically, it removes local state stored in ~/.dnanexus_config/environment.  Does not affect the environment variables currently set in your shell.', prog='dx clearenv')
 parser_clearenv.add_argument('--reset', help='Reset dx environment variables to empty values. Use this to avoid interference between multiple dx sessions when using shell environment variables.',
                              action='store_true')
-parser_clearenv.set_defaults(func=clearenv, interactive=True)
+parser_clearenv.set_defaults(func=clearenv, interactive=False)
```
Comment (psung) [inline src/python/dxpy/scripts/dx.py]: Why is this needed?

File: src/python/dxpy/utils/config.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
+# Copyright (C) 2013-2014 DNAnexus, Inc.
```
Comment (psung) [inline src/python/dxpy/utils/config.py]: 2014 => 2015

**Comment by psung** [comment]
Thanks for taking care of this! Mostly it looks good to me, please see the minor suggestions in the line comments.

I merged latest master and resolved the merge conflicts, and I'm running the integration tests against this branch now.

File: src/python/dxpy/scripts/dx.py (hunk 5 above)
Comment (kislyuk) [inline src/python/dxpy/scripts/dx.py]: Previously, `dx clearenv` would invoke a hardcoded shortcut that would avoid initializing dxpy from the environment. The invocation through the argparse parser would only happen when running `dx sh`. Now we can use the parser for both.

File: src/python/dxpy/scripts/dx.py [hunk 7]
```diff
@@ -427,27 +414,24 @@ def logout(args):
...
         os.environ['DX_PROJECT_CONTEXT_ID'] = project
         if name is not None:
-            set_env_var("DX_PROJECT_CONTEXT_NAME", name)
+            dxpy.config["DX_PROJECT_CONTEXT_NAME"] = name
         if write:
-            write_env_var("DX_PROJECT_CONTEXT_ID", project)
+            dxpy.config.write("DX_PROJECT_CONTEXT_ID", project)
             if name is not None:
-                write_env_var("DX_PROJECT_CONTEXT_NAME", name)
+                dxpy.config.write("DX_PROJECT_CONTEXT_NAME", name)
```
Comment (psung) [inline src/python/dxpy/scripts/dx.py]: Previously these write operations were mostly writing different files, but now they write all the files (resulting in each config file being written multiple times). Consider batching the writes with a single "save" call at the end.

File: src/python/dxpy/utils/config.py (hunk 1 above)
Comment (kislyuk) [inline src/python/dxpy/utils/config.py]: Good point, fixed.

File: src/python/dxpy/utils/config.py (hunk 2 above)
Comment (kislyuk) [inline src/python/dxpy/utils/config.py]: Fixed, thanks.

File: src/python/dxpy/utils/config.py (hunk 3 above)
Comment (kislyuk) [inline src/python/dxpy/utils/config.py]: I'd rather add that later along with lazy loading.

File: src/python/dxpy/utils/config.py (hunk 4 above)
Comment (kislyuk) [inline src/python/dxpy/utils/config.py]: Sounds good to me. Refactored.

File: src/python/dxpy/utils/config.py (hunk 6 above)
Comment (kislyuk) [inline src/python/dxpy/utils/config.py]: Fixed.

File: src/python/dxpy/utils/config.py (hunk 3 above)
Comment (psung) [inline src/python/dxpy/utils/config.py]: SGTM.

File: src/python/dxpy/scripts/dx.py (hunk 7 above)
Comment (kislyuk) [inline src/python/dxpy/scripts/dx.py]: Thanks for pointing this out. I ended up refactoring this block since the logic is simpler now.

File: src/python/dxpy/scripts/dx.py [hunk 8]
```diff
@@ -427,42 +414,29 @@ def logout(args):
...
-        os.environ['DX_PROJECT_CONTEXT_ID'] = project
-        if name is not None:
-            set_env_var("DX_PROJECT_CONTEXT_NAME", name)
-        if write:
-            write_env_var("DX_PROJECT_CONTEXT_ID", project)
-            if name is not None:
-                write_env_var("DX_PROJECT_CONTEXT_NAME", name)
-            else:
-                try:
-                    os.remove(os.path.expanduser('~/.dnanexus_config/DX_PROJECT_CONTEXT_NAME'))
```
Comment (psung) [inline src/python/dxpy/scripts/dx.py]: Thanks for this refactoring, this is much better. Possibly minor point, is it important to distinguish between `DX_PROJECT_CONTEXT_NAME` being None and ""? Before you would remove the corresponding file if the value was set to None, but this is no longer done (with good reason, I think).

**Comment by psung** [comment]
Thanks. Did you ever figure out why the branch tests were failing? Does it have to do with 2669032?

**Comment by kislyuk** [comment]
No. I'm still trying to figure that out. The tests in question pass in isolation. It seems a previous test messes up the environment, or maybe even worse, environment leaks in from another concurrent Jenkins job.

**Comment by kislyuk** [comment]
FYI, https://github.com/dnanexus/dx-toolkit/commit/26690325522b23f1171a5aa313cdd344c9c45c98 is unrelated - it's a fix for a funny problem that I observed on a customer's shell... `LC_CTYPE=UTF-8 dx`

**Comment by kislyuk** [comment]
@psung, after digging around, I found the issue (dxpy.JOB_ID was not being reset). The branchtest passes now. I'm ready to merge.

File: src/python/dxpy/scripts/dx.py (hunk 8 above)
Comment (kislyuk) [inline src/python/dxpy/scripts/dx.py]: They should be treated the same, but you can't set an environment variable to None. Are there any practical differences that you see needed with that distinction in mind?

Comment (psung) [inline src/python/dxpy/scripts/dx.py]: None that I can foresee; I looked through its callers just now too.

File: src/python/dxpy/utils/config.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+            dxpy.set_security_context(json.loads(environ["DX_SECURITY_CONTEXT"]))
+
+        if "DX_JOB_ID" in environ:
+            dxpy.set_job_id(environ["DX_JOB_ID"])
+            dxpy.set_workspace_id(environ.get("DX_WORKSPACE_ID"))
+            dxpy.set_project_context(environ.get("DX_PROJECT_CONTEXT_ID"))
+        elif "DX_PROJECT_CONTEXT_ID" in environ:
+            dxpy.set_job_id(None)
+            dxpy.set_workspace_id(environ["DX_PROJECT_CONTEXT_ID"])
+            dxpy.set_project_context(environ["DX_PROJECT_CONTEXT_ID"])
```
Comment (psung) [inline src/python/dxpy/utils/config.py]: Should we clear these variables if neither DX_JOB_ID nor DX_PROJECT_CONTEXT_ID is set?

**Comment by psung** [comment]
Thanks, nice work! LGTM but please see the one minor suggestion.

File: src/python/dxpy/utils/config.py (hunk 9 above)
Comment (kislyuk) [inline src/python/dxpy/utils/config.py]: Good point, how about now?

**Comment by psung** [comment]
LGTM, thanks!

**Comment by kislyuk** [comment]
Merged in https://github.com/dnanexus/dx-toolkit/commit/952bdf0e146bfbd2bce2c2b00acfec47bca0e304, thanks for the review.
